aws/smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/endpointdiscovery/EndpointDiscoveryValidator.java

```diff
@@ -0,0 +1,224 @@
+                    return Pair.of(operation, infos);
+                })
+                .flatMap(pair -> validateOperation(pair.getLeft(), pair.getRight()).stream())
+                .collect(Collectors.toList());
+    }
+
+    private List<ValidationEvent> validateOperation(OperationShape operation, List<EndpointDiscoveryInfo> infos) {
+        if (infos.isEmpty()) {
+            return Collections.singletonList(error(operation, String.format(
+                    "The operation `%s` is marked with `aws.api#discoveredEndpoint` but is not attached to a "
+                            + "service with the `aws.api#endpointDiscovery` trait.",
+                    operation.getId().toString()
+            )));
+        }
+
+        return infos.stream()
+                .filter(discoveryInfo -> !operation.getErrors().contains(discoveryInfo.getError().getId()))
+                .map(discoveryInfo -> error(operation, String.format(
+                        "The operation `%s` is marked with `aws.api#discoveredEndpoint` and is bound to the service "
+                                + "`%s` but does not have the required error `%s`.",
+                        operation.getId(), discoveryInfo.getService().getId(), discoveryInfo.getError().getId()
+                )))
+                .collect(Collectors.toList());
+    }
+
+    private List<ValidationEvent> validateEndpointOperation(
+            ShapeIndex shapeIndex, OperationIndex opIndex, OperationShape operation
+    ) {
+        List<ValidationEvent> events = new ArrayList<>();
+        opIndex.getInput(operation).ifPresent(input -> {
+            Set<String> memberNames = SetUtils.copyOf(input.getMemberNames());
+            if (!memberNames.equals(SetUtils.of("Operation", "Identifiers"))) {
+                events.add(error(input, String.format(
+                        "Input for endpoint discovery operation `%s` may only have the members Operation and "
+                                + "Identifiers but found: %s",
+                        operation.getId().toString(),
+                        String.join(", ", memberNames)
+                )));
+            }
+
+            input.getMember("Operation")
+                    .flatMap(member -> shapeIndex.getShape(member.getTarget()))
+                    .filter(shape -> !shape.getType().equals(ShapeType.STRING))
+                    .ifPresent(shape -> events.add(error(
+                            shape, "The Operation member of an endpoint discovery operation must be a string")));
+
+            input.getMember("Identifiers")
+                    .map(member -> Pair.of(member, shapeIndex.getShape(member.getTarget())))
+                    .ifPresent(pair -> {
+                        Optional<MapShape> map = pair.getRight().flatMap(Shape::asMapShape);
+                        if (map.isPresent()) {
+                            Optional<Shape> key = shapeIndex.getShape(map.get().getKey().getTarget());
```

Comment:
[The key of a map in Smithy MUST target a `string` by definition](https://awslabs.github.io/smithy/spec/core.html#map), so this can check be removed.